pyro/ops/contract.py

```diff
@@ -487,3 +492,97 @@ def ubersum(equation, *operands, **kwargs):
                 term = term.permute(*map(dims.index, output))
             results.append(term)
     return tuple(results)
+
+
+def naive_ubersum(equation, *operands, **kwargs):
+    """
+    Slow reference implementation of :func:`ubersum`.
+
+    The two implementations should agree whenever both are implemented and
+    :func:`ubersum` does not raise an error.
+    """
+    # Evaluate each output separately.
+    inputs, outputs = equation.split('->')
+    return tuple(_naive_batch_einsum(inputs + '->' + output, *operands, **kwargs)
+                 for output in outputs.split(','))
+
+
+class _DimNormalizer(object):
+    """
+    Object to map batched dims to batches of unbatched dims.
+
+    :param dict dim_to_ordinal: a mapping from contraction dim to the set of
+        batch dims over which the contraction dim is batched.
+    """
+    def __init__(self, dim_to_ordinal):
+        self._plates = {d: tuple(sorted(ordinal)) for d, ordinal in dim_to_ordinal.items()}
+        self._symbols = map(opt_einsum.get_symbol, itertools.count())
+        self._map = {}
+
+    def __call__(self, dim, indices):
+        """
+        Assigns a normalized dim to a base dim, given indices into a batch.
+
+        :param str dim: a dimension to normalize
+        :param dict indices: a mapping from batch dimension to int
+        :return: a normalied dim
+        :rtype: str
+        """
+        plate = self._plates.get(dim, ())
+        index = tuple(indices[d] for d in plate)
+        key = dim, index
+        if key in self._map:
+            return self._map[key]
+        normal_dim = next(self._symbols)
+        self._map[key] = normal_dim
+        return normal_dim
+
+
+def _naive_batch_einsum(equation, *operands, **kwargs):
+    batch_dims = set(kwargs.pop('batch_dims', ''))
+    inputs, output = equation.split('->')
+    inputs = inputs.split(',')
+    output_dims = set(output)
+    keep_dims = output_dims - batch_dims
+    sizes = {dim: size
+             for dims, term in zip(inputs, operands)
+             for dim, size in zip(dims, term.shape)}
+
+    # It is unclear what the desired behavior is in this case.
+    if batch_dims & output_dims:
+        raise NotImplementedError
```

Comment:
It will be nice if `naive_ubersum` could handle cases that are a superset of those handled by `ubersum` (though I suppose it depends on what is the purpose of this reference implementation). One thing we could do is to first call `ubersum` and only go forward with the naive implementation if the case returns a `NotImplementedError`.

Reply:
On second thoughts, that defeats the point of having a reference implementation. Could you explain the difficulty in handling the case where some batch dimensions are not contracted?